include/lbann/utils/serialization/serialize_matrices.hpp

```diff
@@ -0,0 +1,307 @@
+ *  problem. If it is, open an issue and it will be remedied.
+ *
+ *  @tparam ArchiveT (Inferred) The Cereal archive type to use.
+ *  @tparam T (Inferred) The data type of the matrix.
+ *  @param archive The Cereal archive from which the matrix will be read.
+ *  @param mat The target matrix to deserialize.
+ *  @throws lbann::exception The input matrix is already setup as a view.
+ *  @todo Perhaps it's better to throw an exception for these archives?
+ *  @ingroup serialization
+ */
+template <typename ArchiveT, typename T,
+          lbann::utils::WhenTextArchive<ArchiveT> = 1>
+void load(ArchiveT& archive, El::Matrix<T, El::Device::CPU>& mat)
+{
+  LBANN_ASSERT(!mat.Viewing());
+  El::Int height, width;
+  archive(height, width);
+  mat.Resize(height, width);
+}
+
+/** @brief Load a CPU Matrix from a non-text archive.
+ *
+ *  @tparam ArchiveT (Inferred) The Cereal archive type to use.
+ *  @tparam T (Inferred) The data type of the matrix.
+ *  @param archive The Cereal archive from which the matrix will be read.
+ *  @param mat The target matrix to deserialize.
+ *  @throws lbann::exception The input matrix is already setup as a view.
+ *  @ingroup serialization
+ */
+template <typename ArchiveT, typename T,
+          lbann::utils::WhenNotTextArchive<ArchiveT> = 1>
+void load(ArchiveT& archive,
+          El::Matrix<T, El::Device::CPU>& mat)
+{
+  LBANN_ASSERT(!mat.Viewing());
+  El::Int height, width;
+  archive(height, width);
+  mat.Resize(height, width);
+  archive(binary_data(mat.Buffer(),
+                      mat.Height()*mat.Width()*sizeof(T)));
+}
+
+// DistMatrix
+
+template <typename ArchiveT, typename T,
+          lbann::utils::WhenTextArchive<ArchiveT> = 1>
+void save(ArchiveT& ar, El::AbstractDistMatrix<T> const& mat)
+{
+  LBANN_ASSERT(!mat.Viewing());
+
+  if (mat.DistRank() == mat.Root())
+    ar(make_nvp("global_height", mat.Height()),
+       make_nvp("global_width", mat.Width()));
+}
+
+template <typename ArchiveT, typename T,
+          lbann::utils::WhenNotTextArchive<ArchiveT> = 1>
+void save(ArchiveT& ar, El::AbstractDistMatrix<T> const& mat)
+{
+  LBANN_ASSERT(!mat.Viewing());
+
+  using CircMatType =
+    El::DistMatrix<T,El::CIRC,El::CIRC,El::ELEMENT,El::Device::CPU>;
+  CircMatType circ_mat(mat);
+  // Only the root writes to the archive.
+  if (circ_mat.CrossRank() == circ_mat.Root())
+    ar(mat.Height(), mat.Width(), circ_mat.LockedMatrix());
+}
+
+template <typename ArchiveT, typename T,
+          lbann::utils::WhenTextArchive<ArchiveT> = 1>
+void load(ArchiveT& ar, El::AbstractDistMatrix<T>& mat)
+{
+  El::Int height, width;
+  if (mat.DistRank() == mat.Root())
+    ar(height, width);
+  El::mpi::Broadcast(height, mat.Root(), mat.DistComm(),
+                     El::SyncInfo<El::Device::CPU>{});
+  El::mpi::Broadcast(width, mat.Root(), mat.DistComm(),
+                     El::SyncInfo<El::Device::CPU>{});
+  mat.Resize(height, width);
+}
+
+#ifndef DOXYGEN_SHOULD_SKIP_THIS
```

Comment:
I don't understand why this guard is here.

Reply:
It's there so that Doxygen doesn't add documentation for these. But I should remove this since I changed the static dispatch from the original version in which I added that.